### PR TITLE
Reduce noise of e2e coverage comment

### PR DESCRIPTION
**Pull request by piersy**
### Description

This PR aims to make posting the the istanbul coverage comment summary on PRs less noisy.

* Rather than posting a new comment on every commit, (which can be pretty distracting for PRs that are undergoing lots of changes) we now update the coverage comment if it exists already.
* The coverage comment is now expandable and by default only shows overall coverage.

### Tested

See it below, notice it has been edited several times.

### Related issues

#1692 

### Backwards compatibility

No system code was changed.

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit fbea0b049920edfd22c0fa4cb3370dacee31fbe9
```
coverage:  44.5% of statements across all listed packages
coverage:  54.1% of statements in consensus/istanbul
coverage:  45.0% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```
CommentID:{{ checksum pr_url }}

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit bd0160cf7c7fbd33a7369fe9cb1e7c8f2eacf360
```
coverage:  45.0% of statements across all listed packages
coverage:  54.6% of statements in consensus/istanbul
coverage:  45.1% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  65.0% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```
CommentID: {{ checksum pr_url }}

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 560faf7f0478309951a78d13644cb3ec39cb0965
```
coverage:  44.5% of statements across all listed packages
coverage:  54.1% of statements in consensus/istanbul
coverage:  45.0% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.4% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```
CommentID: 4ac5f9cfdc

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 0599f287657f58b4af3d7ea1a65b416ed036942b
```
coverage:  44.9% of statements across all listed packages
coverage:  54.1% of statements in consensus/istanbul
coverage:  45.4% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  64.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```
CommentID: 4ac5f9cfdc

**Review by YoduYodu — COMMENTED** [review]
left some questions

File: .circleci/config.yml [hunk 1]
```diff
@@ -155,27 +156,57 @@ jobs:
...
-              https://api.github.com/repos/celo-org/celo-blockchain/issues/${CIRCLE_PULL_REQUEST##*/}/comments \
-              -d "{\"body\":\"`awk -v ORS='\\\\n' '1' comment`\"}" ;
+            # This expansion deletes everything up to and including the final slash
+            # leaving just the PR ID.
+            PR_ID=${CIRCLE_PULL_REQUEST##*/}
+
+            # This command grabs the list of comments for this PR and selects
+            # the id of the first one that contains the comment marker in its
+            # body, jq returns 'null' when it doesn't find anything.
+            COMMENT_ID=`curl https://api.github.com/repos/celo-org/celo-blockchain/issues/${PR_ID}/comments | jq "[.[] | select(.body | contains(\"${COMMENT_MARKER}\"))][0]| .id"`
```
[the remark this answers is not in the document]
Comment (piersy) [reply .circleci/config.yml:186]: Its a little strange but that is how it works - https://docs.github.com/en/rest/reference/pulls#review-comments
Issue comments basically cover comments that do not reference a specific line of code, whether they are on an issue or a PR.

File: .circleci/config.yml [hunk 2]
```diff
@@ -155,27 +156,57 @@ jobs:
...
+
+            # Determine if we are posting  a new comment or patching an
+            # existing one.
+            if [[ $COMMENT_ID == "null" ]]; then
+              # There was no previous comment
+              CURL_VERB=POST
+              URL=https://api.github.com/repos/celo-org/celo-blockchain/issues/${PR_ID}/comments
```
[the remark this answers is not in the document]
Comment (piersy) [reply .circleci/config.yml:193]: So yeah, its meant to be issues.

File: .circleci/config.yml [hunk 3]
```diff
@@ -155,27 +156,57 @@ jobs:
...
+            else
+              # We are updating a previous comment
+              CURL_VERB=PATCH
+              URL=https://api.github.com/repos/celo-org/celo-blockchain/issues/comments/${COMMENT_ID}
```
[the remark this answers is not in the document]
Comment (piersy) [reply .circleci/config.yml:197]: Same again.

File: .circleci/config.yml [hunk 4]
```diff
@@ -76,6 +76,7 @@ jobs:
             - <<parameters.cache-key>>
       - attach_workspace:
           at: ~/repos
+      - run: jq -h
```
[the remark this answers is not in the document]
Comment (piersy) [reply .circleci/config.yml:79]: Thanks! This is left over debug, it should have been removed.

**Review by YoduYodu — APPROVED** [review]
LGTM

**Pull request merged**